File: Leetcode/222-Count-Complete-Tree-Nodes/CountCompleteTreeNodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
	h := height(root)
	if h < 0 {
		return 0
	} else {
		if height(root.Right) == h-1 {
			return (1 << uint(h)) + countNodes(root.Right)
		} else {
			return (1 << uint(h-1)) + countNodes(root.Left)
		}
	}
}

func height(node *TreeNode) int {
	if node == nil {
		return -1
	} else {
		return 1 + height(node.Left)
	}
}

File: Leetcode/222-Count-Complete-Tree-Nodes/CountCompleteTreeNodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function(root) {
    var h = height(root)
    return h < 0? 0 :
        height(root.right) == h-1?
            (1<<h) + countNodes(root.right) :
            (1<<h-1) + countNodes(root.left)
};

var height = function(node) {
    return node == null? -1 : 1+height(node.left)
};
